### PR TITLE
unify tuple creation intrinsics

### PR DESCRIPTION
**Pull request by ARF1**
This PR unifies two tuple creation intrinsics, generalizes them and makes them safe to use.

### CODE REVIEW

**Comment by esc** [comment]
@ARF1 thanks for submitting this, I have queued it for review.

File: numba/tests/test_tuples.py
Comment (stuartarchibald) [inline numba/tests/test_tuples.py]: yes

File: numba/tests/test_unsafe_intrinsics.py
Comment (stuartarchibald) [inline numba/tests/test_unsafe_intrinsics.py]: yes

File: numba/unsafe/ndarray.py
Comment (stuartarchibald) [inline numba/unsafe/ndarray.py]: I'm not sure it adds any value?

@sklam do you recall how prevalent this is in external code?

File: numba/targets/tupleobj.py
Comment (esc) [reply numba/targets/tupleobj.py]: Yeah, https://github.com/numba/numba/blob/master/numba/listobject.py#L955 <--- here is one of those. IIRC the typed-list now has `dtype` as an alias. Refactoring would involve adding a utility somewhere, maybe close to iterator/iterable code or in the `types.utils` module if it exists. Something like `dtype_from_iterable`?

File: numba/targets/tupleobj.py
Comment (ARF1) [reply numba/targets/tupleobj.py]: @stuartarchibald I did not put it in `numba.unsafe` because I thought it was no longer "unsafe". While it does use `tuple_setitem` which is unsafe, I though I caught the cases that would make use of `tuple_setitem` unsafe. Did I overlook something? In any case, I am happy to move it.

File: numba/unsafe/ndarray.py
Comment (ARF1) [reply numba/unsafe/ndarray.py]: This might not be representative but I found [two projects on github](https://github.com/search?utf8=%E2%9C%93&q=%22to_fixed_tuple%22+-filename%3A%22test_unsafe_intrinsics.py%22+-filename%3A%22ndarray.py%22+-filename%3A%22devicearray.py%22+extension%3Apy&type=Code&ref=advsearch&l=&l=) that use this function in four places.

File: numba/targets/tupleobj.py
Comment (ARF1) [reply numba/targets/tupleobj.py]: Done.

**Comment by ARF1** [comment]
@stuartarchibald Thanks for the review. I resolved what I could for now, without the comments you requested from others.

File: numba/targets/tupleobj.py
Comment (esc) [reply numba/targets/tupleobj.py]: Or place it in the same place that all the overloading/extending utils go, such as `_is_nonelike`. I am not sure if came to a conclusion where this module should live. I would propose to leave it like this for now however.

File: numba/tests/test_tuples.py
Comment (esc) [inline numba/tests/test_tuples.py]: might be worth adding a simple list comprehension, a typed-list and typed-dict to this.

File: numba/targets/tupleobj.py
Comment (sklam) [inline numba/targets/tupleobj.py]: I think this has to be a `.get_constant_null` to zero-fill in case the tuple contains refcounted types.

File: numba/tests/test_tuples.py
Comment (esc) [inline numba/tests/test_tuples.py]: might be worth adding a typed-list and typed-dict to this.

File: numba/tests/test_tuples.py
Comment (esc) [inline numba/tests/test_tuples.py]: Not sure about this test name, was it perhaps inadvertently copied from the now removed tests?

File: numba/unsafe/ndarray.py
Comment (esc) [reply numba/unsafe/ndarray.py]: Just to clarify, this is the second intrinsic that will be removed?

Comment (ARF1) [reply numba/unsafe/ndarray.py]: @esc Yes, this is the second intrinsic.

File: numba/targets/arraymath.py
Comment (ARF1) [inline numba/targets/arraymath.py]: @esc This is the "first" intrinsic.

File: numba/targets/tupleobj.py
Comment (ARF1) [reply numba/targets/tupleobj.py]: @sklam I am sorry but I have no idea. I will need some definite guidance here. You want me to make the change just to be safe or is there a way I can test its necessity?

File: numba/tests/test_tuples.py
Comment (ARF1) [reply numba/tests/test_tuples.py]: @esc Good point. I will do that. I am still a bit fuzzy though on typed lists and reflected lists: Are not all lists created in nopython mode typed? If yes, this would already be a typed-list.

File: numba/tests/test_tuples.py
Comment (ARF1) [reply numba/tests/test_tuples.py]: @esc It was indeed copy-pasted. I originally kept it because it contains some test-cases that I did not test explicitly elsewhere. I will copy its contents to a more appropriate test case.

File: numba/unsafe/ndarray.py
Comment (esc) [reply numba/unsafe/ndarray.py]: ✅

File: numba/targets/arraymath.py
Comment (esc) [reply numba/targets/arraymath.py]: thanks!

File: numba/tests/test_tuples.py
Comment (esc) [reply numba/tests/test_tuples.py]: There are two types of list currently available in Numba, the so-called *reflected-list* and the *typed-list*.  You can read more about them and their differences here: http://numba.pydata.org/numba-doc/latest/reference/pysupported.html#list - let me know if the documentation leaves any unanswered questions.

Comment (ARF1) [reply numba/tests/test_tuples.py]: @esc Thanks. I thought this was just semantics to help the transition from python space. I did not realise they are internally different implementations.

Good call testing the overlooked dict types in particular: they were failing...

File: numba/tests/test_tuples.py
Comment (ARF1) [reply numba/tests/test_tuples.py]: Done.

File: numba/targets/tupleobj.py
Comment (ARF1) [reply numba/targets/tupleobj.py]: After fixing some problems with `dict` and typed list, the cascade has got even longer. Now I also have `key_type` and `item_type`... Can this be nevertheless be marked as resolved?

File: numba/unsafe/ndarray.py
Comment (ARF1) [reply numba/unsafe/ndarray.py]: I am going to go out on a limb and say that in light of the limited external usage, it is probably reasonable to delete this in the interest of code cleanliness. I am happy to give the two projects using this on github a heads-up if you agree. I have deleted it for now. Please yell if you want it restored.

File: numba/targets/tupleobj.py
Comment (esc) [reply numba/targets/tupleobj.py]: I would tend to answer yes as refactoring of this should be taken care of in a subsequent PR

**Review by esc — DISMISSED** [review]
I have given it another review and was unable to find anything else to comment on, so, green light from my end.

**Comment by esc** [comment]
@ARF1 the next step, however, would be to merge in current master. We refactored the world with https://github.com/numba/numba/pull/5197 -- so expect merge conflicts 🥶

**Comment by ARF1** [comment]
@esc I was planning to rebase before the merge to clean up the history. I am just waiting until the remaining issues are resolved so as not to annoy anybody...

File: numba/targets/tupleobj.py
Comment (ARF1) [reply numba/targets/tupleobj.py]: @sklam I looked into this myself and to be honest I cannot follow your reasoning. If I understand correctly `get_constant_undef()` does something akin to pure allocation while `get_constant_null()` does additinally zero fill, right?

Since all slots in the tuple will be filled any zero-filling is redundant. Anything not filling the tuple completely would raise an error.

The null-initialization seems to get optimized away. The resulting llvm is identical whether I use `get_constant_undef()` or `get_constant_null()`.

That said, you were specifically mentioning refcounted types as the possible problem. I do not know which they are and thus could not test this case specifically.

Do you want me to just use `get_constant_null()` to be safe and be done with it? After all it does not seem to hurt in the case of non-refcounted types.

---
This is the test-code I used:

```python
from numba import njit
from numba.cpython.tupleobj import make_tuple

@njit
def test_tuple_creation():
    return make_tuple(5, range(5))

test_tuple_creation()
print(test_tuple_creation.inspect_llvm(test_tuple_creation.signatures[0]))
```

**Comment by ARF1** [comment]
@stuartarchibald From my side this is ready to merge. Unless you still want me to move it to unsafe? (See remaining open conversation above.)

**Comment by stuartarchibald** [comment]
@ARF1 thanks, I haven't reviewed the code yet, but just tried out a few things, these all seem to break in various ways:
```python
from numba import njit
import numpy as np
from numba.cpython.tupleobj import make_tuple

@njit
def foo():
    d = make_tuple(-1, (1, 2, 3))
    d = make_tuple(3, (1, 1j, 'a'))
    d = make_tuple(3, (1, 2, 3))
    d = make_tuple(3, "123")
    d = make_tuple(2, [np.zeros(4,), np.zeros(6,)])
    def gen():
        for x in range(3):
            yield x
    d = make_tuple(2, gen)

print(foo())
```
I think that some should work and some are doing the wrong thing. In general, given how intrinsics are "dangerous" but incredibly useful it'd be good to spend a bit of time investigating if it can be made more robust. Thanks.

**Comment by ARF1** [comment]
@stuartarchibald Thanks for taking the time. Because of the "waiting on the author" tag I thought you had finished your review and only the safe/unsafe question was open. Sorry for the confusion.

In the meantime I will get to work on fixing at least the issues you highlighted.

**Comment by ARF1** [comment]
@stuartarchibald I fixed the issues you found but in the process I created a memory leak with which I could use help when you have some time.

The basic process is now:

1. Convert input to a list of the specified type, catching any errors. (This takes care of creating any new ref-counted elements as required by `make_tuple(3, '123')`.)
2. Then I am copying the already allocated list elements into a tuple.

I suspect the memory leak may have possibly two causes:
1. I am fairly certain I need to destroy the temporary list after I am done with it but without releasing its ref-counted elements. How do I do that? 
2. I suspect I will still have a "memory leak" simply because `make_tuple(3, '123')` allocates new ref-counted elements. (I.e. three new UnicodeTypes.) Is this right? If yes, I am not sure how to test that the correct amount of memory "leaks".

**Comment by ARF1** [comment]
@stuartarchibald Thanks to @sklam fixing #5338 this is now working and is "ready for review".

@sklam There are a test error that stems from master. (`numba.tests.test_nrt.TestRefCtPruning.test_refct_pruning_with_branches`) Just in case you have not seen it yet yourself.

**Comment by ARF1** [comment]
Force pushed without changes to retrigger CI after #5345 was fixed.

File: numba/cpython/tupleobj.py
Comment (stuartarchibald) [inline numba/cpython/tupleobj.py:536]: I'm wondering if this approach could be simplified. At compile time the size of the tuple is known, so can be allocated via an `@intrinsic`, subsequently all writes can be done via an intrinsic too, and the driving code be done in a `njit` function? I'm struggling to understand the need for a list buffer too?

**Comment by ulupo** [comment]
Is there a status update on this?  If I understood correctly there are some reviewer comments which still need to be addressed by the author.  This would be a very useful feature for my applications and (anecdotal as the following may be) I think the lack of ability to use the `tuple` constructor periodically trips over some users, see e.g. #2771.

**Comment by ziofil** [comment]
bump

**Comment by majra20** [comment]
What's the status of this pr?

**Comment by ziofil** [comment]
bump 🥺

**Comment by gmarkall** [comment]
It seems that there's interest in this PR but the original author is no longer participating. @ziofil @majra20 @ulupo would any of you be interested in taking it to completion? (possibly as a new PR based on the work of this one so far)

**Comment by gmarkall** [comment]
I haven't time to look deeper into this PR, but I also wonder whether there's any overlap / interaction with #8029 (cc @dlee992 in case this is of any interest)

**Comment by dlee992** [comment]
> I haven't time to look deeper into this PR, but I also wonder whether there's any overlap / interaction with #8029 (cc @dlee992 in case this is of any interest)

From a quick look at this PR, it seems unrelated with #8029. I think it isn't complicated, maybe I will take it to completion after finishing my current PRs. But I also see some "related" PRs, e.g., #7270 and #7401, are ready for review, but after more than half a year, there is no update on these PRs.

**Comment by ulupo** [comment]
> @ziofil @majra20 @ulupo would any of you be interested in taking it to completion?

Thanks for the suggestion @gmarkall.  Unfortunately, I don't have the skills required to contribute to this PR at this time.

**Comment by gmarkall** [comment]
Thanks for responding on this @ulupo, @ziofil and @dlee992 (out of band). To try and move things along a little here, I've merged main and done a couple of necessary fixups. There are a couple of issues outstanding from the CUDA side that I now notice:

- The use of the list buffer causes an issue for CUDA because creating a typed list and extending it is not supported there.
- The changes now result in the import of `numba.cpython.tupleobj` as a side effect of importing CUDA, resulting in a fail in `numba.cuda.tests.nocuda.test_import.TestImport.test_no_impl_import`. It may be that this is OK and the test just needs fixing to not ban the import of the `tupleobj` module.

I'll run the gpuCI so that the CUDA-related issue is clear.

**Comment by gmarkall** [comment]
gpuci run tests

**Comment by gmarkall** [comment]
gpuci run tests

**Comment by gmarkall** [comment]
Example of errors in the CUDA target from https://gpuci.gpuopenanalytics.com/job/numba/job/numba/job/prb/job/numba-prb/201/CUDA_TOOLKIT_VER=11.5,CUDA_VER=11.2,LINUX_VER=ubuntu18.04,PYTHON_VER=3.8,RAPIDS_VER=21.12/console:

```
10:27:10 numba.core.errors.TypingError: Failed in cuda mode pipeline (step: native lowering)
10:27:10 Failed in nopython mode pipeline (step: nopython frontend)
10:27:10 - Resolution failure for literal arguments:
10:27:10 No implementation of function Function(<function impl_append at 0x7f8989eac3a0>) found for signature:
10:27:10 
10:27:10  >>> impl_append(ListType[undefined], int64)
10:27:10 
10:27:10 There are 2 candidate implementations:
10:27:10       - Of which 2 did not match due to:
10:27:10       Overload in function 'impl_append': File: numba/typed/listobject.py: Line 592.
10:27:10         With argument(s): '(ListType[undefined], int64)':
10:27:10        Rejected as the implementation raised a specific error:
10:27:10          NumbaRuntimeError: Failed in nopython mode pipeline (step: native lowering)
10:27:10        NRT required but not enabled
10:27:10        During: lowering "casteditem = call $2load_global.0(item, $6load_deref.2, func=$2load_global.0, args=[Var(item, listobject.py:600), Var($6load_deref.2, listobject.py:600)], kws=(), vararg=None, varkwarg=None, target=None)" at /workspace/numba/typed/listobject.py (600)
10:27:10   raised from /workspace/numba/core/runtime/context.py:17
10:27:10 
10:27:10 - Resolution failure for non-literal arguments:
10:27:10 None
10:27:10 
10:27:10 During: resolving callee type: BoundFunction((<class 'numba.core.types.containers.ListType'>, 'append') for ListType[undefined])
10:27:10 During: typing of call at /workspace/numba/cpython/tupleobj.py (530)
10:27:10 
10:27:10 
10:27:10 File "numba/cpython/tupleobj.py", line 530:
10:27:10         def convert_to_list(iterable):
10:27:10             <source elided>
10:27:10             for i in range(tuple_size):
10:27:10                 out.append(iterable[i])
10:27:10                 ^
10:27:10 
10:27:10 During: lowering "$186call_function.6 = call $176load_global.1($178load_deref.2, $184binary_subscr.5, func=$176load_global.1, args=[Var($178load_deref.2, devicearray.py:542), Var($184binary_subscr.5, devicearray.py:542)], kws=(), vararg=None, varkwarg=None, target=None)" at /workspace/numba/cuda/cudadrv/devicearray.py (542)
10:27:10 
10:27:10 ======================================================================
10:27:10 FAIL: test_no_impl_import (numba.cuda.tests.nocuda.test_import.TestImport)
10:27:10 Tests that importing cuda doesn't trigger the import of modules
10:27:10 ----------------------------------------------------------------------
10:27:10 Traceback (most recent call last):
10:27:10   File "/workspace/numba/cuda/tests/nocuda/test_import.py", line 45, in test_no_impl_import
10:27:10     self.assertFalse(unexpected, "some modules unexpectedly imported")
10:27:10 AssertionError: {'numba.cpython.tupleobj'} is not false : some modules unexpectedly imported
10:27:10
```

(note the unexpected import, and the `NRT required but not enabled` message due to the use of the typed list)

**Comment by esc** [comment]
@gmarkall may I kindly ask: what is the state of this?

**Comment by gmarkall** [comment]
gpuci run tests

**Comment by gmarkall** [comment]
gpuci run tests

**Comment by gmarkall** [comment]
Assuming tests are OK (except gpuCI) after the latest merge, the to-do here is:

- Fix the CUDA issues outlined in https://github.com/numba/numba/pull/5169#issuecomment-1148400650
- Check whether things can be handled in a similar manner as suggested by @stuartarchibald in https://github.com/numba/numba/pull/5169#discussion_r404034198

**Comment by gmarkall** [comment]
gpuci run tests

**Comment by gmarkall** [comment]
gpuci run tests

**Comment by gmarkall** [comment]
This is still on my to-do list.

**Pull request closed without merging**
